pkg/aws/awscontrolplane/admit_awscontrolplane.go

```diff
@@ -0,0 +1,150 @@
+package awscontrolplane
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	infrastructurev1alpha2 "github.com/giantswarm/apiextensions/pkg/apis/infrastructure/v1alpha2"
+	releasev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/release/v1alpha1"
+	"github.com/giantswarm/apiextensions/pkg/label"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+	restclient "k8s.io/client-go/rest"
+	apiv1alpha2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+
+	"github.com/giantswarm/admission-controller/pkg/admission"
+	"github.com/giantswarm/admission-controller/pkg/aws"
+)
+
+type Admitter struct {
+	k8sClient              k8sclient.Interface
+	validAvailabilityZones []string
+	logger                 micrologger.Logger
+}
+
+type Config struct {
+	ValidAvailabilityZones string
+	Logger                 micrologger.Logger
+}
+
+// var (
+//  awsControlPlaneResource = metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "awscontrolplane"}
+// )
+
+func NewAdmitter(config Config) (*Admitter, error) {
+	var k8sClient k8sclient.Interface
+	{
+		restConfig, err := restclient.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("failed to load key kubeconfig: %v", err)
+		}
+		c := k8sclient.ClientsConfig{
+			SchemeBuilder: k8sclient.SchemeBuilder{
+				apiv1alpha2.AddToScheme,
+				infrastructurev1alpha2.AddToScheme,
+				releasev1alpha1.AddToScheme,
+			},
+			Logger: config.Logger,
+
+			RestConfig: restConfig,
+		}
+
+		k8sClient, err = k8sclient.NewClients(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var availabilityZones []string = strings.Split(config.ValidAvailabilityZones, ",")
+	admitter := &Admitter{
+		k8sClient:              k8sClient,
+		validAvailabilityZones: availabilityZones,
+		logger:                 config.Logger,
+	}
+
+	return admitter, nil
+}
+
+func (a *Admitter) Admit(request *v1beta1.AdmissionRequest) ([]admission.PatchOperation, error) {
+	awsControlPlaneCR := &infrastructurev1alpha2.AWSControlPlane{}
+	if _, _, err := admission.Deserializer.Decode(request.Object.Raw, nil, awsControlPlaneCR); err != nil {
+		return nil, microerror.Maskf(aws.ParsingFailedError, "unable to parse awscontrol plane: %v", err)
+	}
+	releaseVersion, err := releaseVersion(awsControlPlaneCR)
+	if err != nil {
+		return nil, microerror.Maskf(aws.ParsingFailedError, "unable to parse release version from AWSControlPlane")
+	}
+
+	var result []admission.PatchOperation
+	// Trigger defaulting of the master availability zones
+	if awsControlPlaneCR.Spec.AvailabilityZones == nil && aws.IsHAVersion(releaseVersion) {
```

Comment:
A best practice is to return early and safe the additional indentation for the majority of the code below.

Reply:
Good point. I'll do it like that. Although it may change when we add more defaulting and validation later.

Reply:
In anticipation you could think about a pattern that makes it easier to deal with. The [`ruleengine`](https://github.com/giantswarm/ruleengine) package made one attempt to ease such a pain.